src/ucx_plugin.c

```diff
@@ -632,50 +639,74 @@ ncclResult_t ucx_send_check(ucx_send_comm_t *comm) {
   return ncclSuccess;
 }
 
-ncclResult_t ucx_recv_check(ucx_recv_comm_t *comm) {
-  ucp_address_t *my_addr;
-  size_t        local_addr_len;
-  size_t        msg_len;
-
-  if (comm->connect_req == NULL) {
-    NCCLCHECK(ucx_worker_get_netaddress(comm->worker, &my_addr, &local_addr_len));
-    nccl_ucx_add_ep(comm->worker, &comm->sock);
-    msg_len = sizeof(connect_msg_t) + local_addr_len;
-    comm->msg = calloc(1, msg_len);
-    comm->msg->addr_len = local_addr_len;
-    memcpy(comm->msg + 1, my_addr, local_addr_len);
-    comm->connect_req = ucp_tag_send_nb(comm->ep, comm->msg, msg_len,
-                                        ucp_dt_make_contig(1), comm->ctag,
-                                        send_handler);
-
-    if (UCS_PTR_IS_ERR(comm->connect_req)) {
-      WARN("Unable to send connect message");
-      return ncclSystemError;
-    } else if (comm->connect_req == NULL){
-      comm->ready = 1;
-      free(comm->msg);
-    }
+static void ucx_recv_set_ready(ucx_comm_t *comm) {
+  free(comm->msg);
+  comm->msg   = NULL;
+  comm->ready = 1;
+}
 
-    free(my_addr);
-  } else {
-    if (comm->connect_req->completed == 1) {
-      comm->ready = 1;
-      comm->connect_req->completed = 0;
-      ucp_request_free(comm->connect_req);
-      free(comm->msg);
-    }
-    else {
-      ucp_worker_progress(comm->worker);
-    }
+static void check_handler(void *request, ucs_status_t status, void *user_data) {
+  assert(status == UCS_OK);
+  ucx_recv_set_ready((ucx_comm_t*)user_data);
+  ucp_request_free(request);
+}
+
+ncclResult_t ucx_recv_check(ucx_comm_t *comm) {
+  ucp_request_param_t params;
+  ucp_address_t       *my_addr;
+  size_t              local_addr_len;
+  size_t              msg_len;
+
+  if (comm->connect_req != NULL) {
+    goto done;
+  }
+
+  NCCLCHECK(ucx_worker_get_netaddress(comm->worker, &my_addr, &local_addr_len));
+  nccl_ucx_add_ep(comm->worker, &comm->sock);
+
+  msg_len             = sizeof(connect_msg_t) + local_addr_len;
+  comm->msg           = calloc(1, msg_len);
+  comm->msg->addr_len = local_addr_len;
+  memcpy(comm->msg + 1, my_addr, local_addr_len);
+
+  params.op_attr_mask = UCP_OP_ATTR_FIELD_CALLBACK |
+                        UCP_OP_ATTR_FIELD_USER_DATA |
+                        UCP_OP_ATTR_FIELD_DATATYPE;
+  params.cb.send      = check_handler;
+  params.user_data    = comm;
+  params.datatype     = ucp_dt_make_contig(1);
```

Comment:
we can avoid passing datatype here as well
(`ucp_dt_make_contig(1)` is a default option when `UCP_OP_ATTR_FIELD_DATATYPE` is not passed)